### DISCUSSION TITLE
Ability to update file(s) in an already existing partition

### DISCUSSION THREAD

**Discussion by obbardc** [rauc/rauc#1446] [Q&A]
Is there an ability to update a file/set of files in an already existing partition?

The requirement is to upgrade a Kernel UKI in a vfat boot partition but leaving the other files alone.

As a quick workaround, in a `post-install` script for the rootfs I mounted the partition, copied the file from the bundle then unmounted the partition.

**Comment by jluebbe** [comment]
> Is there an ability to update a file/set of files in an already existing partition?
> 
> The requirement is to upgrade a Kernel UKI in a vfat boot partition but leaving the other files alone.

RAUC doesn't have a built-in way to do this, as writing to a vfat partition is not atomic, so it cannot be as robust as other ways. So far, we didn't have projects where that downside was acceptable, so we've not implemented it. :)

> As a quick workaround, in a `post-install` script for the rootfs I mounted the partition, copied the file from the bundle then unmounted the partition.

That would be my recommended workaround as well, if you're fine with the additional risk during writes to the vfat partition.

One alternative is to have separate boot partitions per rootfs and to update the EFI boot variables to point to those. That's basically what the EFI bootloader backend does: https://rauc.readthedocs.io/en/latest/integration.html#efi. Whether EFI variable updates are atomic depends on your UEFI implementation, though.

It would be nice to have a way to mount fat filesystems in a way that would guarantee that existing files are not touched and changes are safe against power cuts and crashes, but I've not seen such an implementation yet. I've got some ideas how that could be done (perhaps explicit superblock updates inside the vfat driver or keeping a separate RW copy and switching that via device mapper), but no concrete need/time for that yet.
